src/masternodeman.cpp

```diff
@@ -644,6 +671,401 @@ void CMasternodeMan::ProcessMessage(CNode* pfrom, std::string& strCommand, CData
         }
         // smth weird happen - someone asked us for vin we have no idea about?
         LogPrint("masternode", "CMasternodeMan::ProcessMessage -- DSEG -- No invs sent to peer %d\n", pfrom->id);
+
+    } else if (strCommand == NetMsgType::MNVERIFY) { // Masternode Verify
+
+        CMasternodeVerification mnv;
+        vRecv >> mnv;
+
+        if(mnv.vchSig1.empty()) {
+            // CASE 1: someone asked me to verify myself /IP we are using/
+            SendVerifyReply(pfrom, mnv);
+        } else if (mnv.vchSig2.empty()) {
+            // CASE 2: we _probably_ got verification we requested from some masternode
+            ProcessVerifyReply(pfrom, mnv);
+        } else {
+            // CASE 3: we _probably_ got verification broadcast signed by some masternode which verified another one
+            ProcessVerifyBroadcast(pfrom, mnv);
+        }
+    }
+}
+
+void CMasternodeMan::DoFullVerificationStep()
+{
+    if(activeMasternode.vin == CTxIn()) return;
+
+    int nCountInitial = -1;
+    int nCountMax = std::max(10, (int)vMasternodes.size() / 100); // verify at least 10 masternode at once but at most 1% of all known masternodes
+    int nCount = nCountInitial;
+
+    std::vector<std::pair<int, CMasternode> > vecMasternodeRanks = GetMasternodeRanks(pCurrentBlockIndex->nHeight - 1, MIN_POSE_PROTO_VERSION);
+    std::vector<CMasternode*> vSortedByAddr;
+
+    LOCK(cs);
+
+    BOOST_FOREACH(CMasternode& mn, vMasternodes) {
+        vSortedByAddr.push_back(&mn);
+    }
+
+    // send verify requests to up to nCountMax masternodes starting next to self (by rank)
+    // but only if we are in top MAX_POSE_RANK
```

Comment:
IMO,  In case of Verify being send by first 10 in rank, send to next 10 after its rank. So, the MN at 10th rank will send out 10 request (to 11th - 20th) and will get requests from 9 MNs (1st - 9th). I think it's a bit unfair. 
How about send requests to masternodes starting next to (MAX_POSE_RANK * own rank) ? So, MN should request/reply verify only once per block.